# I -  Exploratory data analysis

As in different data projects, we'll first start diving into the data and build up our first intuitions.

In this section, we'll be doing four things. 

- Data extraction : we'll load the dataset and have a first look at it. 
- Cleaning : we'll fill in missing values.
- Plotting : we'll create some interesting charts that'll (hopefully) spot correlations and hidden insights out of the data.
- Assumptions : we'll formulate hypotheses from the charts.

We tweak the style of this notebook a little bit to have centered plots.

In [1]:
from IPython.core.display import HTML
HTML("""
<style>
.output_png {
    display: table-cell;
    text-align: center;
    vertical-align: middle;
}
</style>
""");

We import the useful libraries.

In [2]:
%matplotlib inline

import warnings
warnings.filterwarnings('ignore')
warnings.filterwarnings('ignore', category=DeprecationWarning)

import pandas as pd
pd.options.display.max_columns = 100

from matplotlib import pyplot as plt
import numpy as np

import seaborn as sns

import pylab as plot
params = { 
    'axes.labelsize': "large",
    'xtick.labelsize': 'x-large',
    'legend.fontsize': 20,
    'figure.dpi': 150,
    'figure.figsize': [25, 7]
}
plot.rcParams.update(params)

Two datasets are available: a training set and a test set.
We'll be using the training set to build our predictive model and the testing set to score it and generate an output file to submit on the Kaggle evaluation system.

We'll see how this procedure is done at the end of this post.

Now let's start by loading the training set.

In [14]:
data = pd.read_csv('data/train.csv')
test = pd.read_csv('data/test.csv')

In [15]:
print (data.shape)
print (test.shape)

(95662, 16)
(45019, 15)


We have:

- 95662 rows
- 16 columns

Pandas allows you to have a sneak peak at your data.

In [16]:
data.head()

TransactionId         BatchId       AccountId       SubscriptionId  \
0  TransactionId_76871   BatchId_36123  AccountId_3957   SubscriptionId_887   
1  TransactionId_73770   BatchId_15642  AccountId_4841  SubscriptionId_3829   
2  TransactionId_26203   BatchId_53941  AccountId_4229   SubscriptionId_222   
3    TransactionId_380  BatchId_102363   AccountId_648  SubscriptionId_2185   
4  TransactionId_28195   BatchId_38780  AccountId_4841  SubscriptionId_3829   

        CustomerId CurrencyCode  CountryCode    ProviderId     ProductId  \
0  CustomerId_4406          UGX          256  ProviderId_6  ProductId_10   
1  CustomerId_4406          UGX          256  ProviderId_4   ProductId_6   
2  CustomerId_4683          UGX          256  ProviderId_6   ProductId_1   
3   CustomerId_988          UGX          256  ProviderId_1  ProductId_21   
4   CustomerId_988          UGX          256  ProviderId_4   ProductId_6   

      ProductCategory    ChannelId   Amount  Value  TransactionStartTime  \
0             airtime  ChannelId_3   1000.0   1000  2018-11-15T02:18:49Z   
1  financial_services  ChannelId_2    -20.0     20  2018-11-15T02:19:08Z   
2             airtime  ChannelId_3    500.0    500  2018-11-15T02:44:21Z   
3        utility_bill  ChannelId_3  20000.0  21800  2018-11-15T03:32:55Z   
4  financial_services  ChannelId_2   -644.0    644  2018-11-15T03:34:21Z   

   PricingStrategy  FraudResult  
0                2            0  
1                2            0  
2                2            0  
3                2            0  
4                2            0

- Spliting the TransactionStartTime into year, months, days, weeks and time for better analysis
- dropping the TransactionStartTime column its irrelevant

In [17]:
data['TransactionStartTime'] = pd.to_datetime(data['TransactionStartTime'])
data['year'] = data['TransactionStartTime'].dt.year
data['month'] = data['TransactionStartTime'].dt.month
data['day'] = data['TransactionStartTime'].dt.day
data['week'] = data['TransactionStartTime'].dt.dayofweek
data['time'] = data['TransactionStartTime'].dt.time
data.drop('TransactionStartTime', axis=1, inplace=True)

In [18]:
data.head()

TransactionId         BatchId       AccountId       SubscriptionId  \
0  TransactionId_76871   BatchId_36123  AccountId_3957   SubscriptionId_887   
1  TransactionId_73770   BatchId_15642  AccountId_4841  SubscriptionId_3829   
2  TransactionId_26203   BatchId_53941  AccountId_4229   SubscriptionId_222   
3    TransactionId_380  BatchId_102363   AccountId_648  SubscriptionId_2185   
4  TransactionId_28195   BatchId_38780  AccountId_4841  SubscriptionId_3829   

        CustomerId CurrencyCode  CountryCode    ProviderId     ProductId  \
0  CustomerId_4406          UGX          256  ProviderId_6  ProductId_10   
1  CustomerId_4406          UGX          256  ProviderId_4   ProductId_6   
2  CustomerId_4683          UGX          256  ProviderId_6   ProductId_1   
3   CustomerId_988          UGX          256  ProviderId_1  ProductId_21   
4   CustomerId_988          UGX          256  ProviderId_4   ProductId_6   

      ProductCategory    ChannelId   Amount  Value  PricingStrategy  \
0             airtime  ChannelId_3   1000.0   1000                2   
1  financial_services  ChannelId_2    -20.0     20                2   
2             airtime  ChannelId_3    500.0    500                2   
3        utility_bill  ChannelId_3  20000.0  21800                2   
4  financial_services  ChannelId_2   -644.0    644                2   

   FraudResult  year  month  day  week      time  
0            0  2018     11   15     3  02:18:49  
1            0  2018     11   15     3  02:19:08  
2            0  2018     11   15     3  02:44:21  
3            0  2018     11   15     3  03:32:55  
4            0  2018     11   15     3  03:34:21

- Getting to know the total amout  pof customers we have and number of transactions by each customers


In [24]:
print('We have ',data.CustomerId.nunique(),'CustomerID  \n and also we have ', data.TransactionId.nunique(), '  TransactionId \n which summounts that each customers made more than on transactions')

We have  3742 CustomerID  
 and also we have  95662   TransactionId 
 which summounts that each customers made more than on transactions
